datasets/program_executor.py

```diff
@@ -0,0 +1,28 @@
+from abc import ABC
+
+from datasets.context import Context
+
+
+class ProgramExecutor(ABC):
+    """
+    Class to access information about the ML program currently being executed.
+    """
+
+    @property
```

Comment:
I'm guessing we want these to be `abstract` properties? Looks like you can combine these with `@abstractmethod` to enforce this if that's the behavior we want to codify 😄  https://stackoverflow.com/a/48710068/724251